pilot/model/context.go

```diff
@@ -93,11 +93,23 @@ func ParseServiceNode(s string) (Node, error) {
 	out := Node{}
 
 	if len(parts) != 4 {
-		return out, errors.New("missing parts in the service node")
+		return out, fmt.Errorf("missing parts in the service node %q", s)
 	}
 
 	out.Type = NodeType(parts[0])
+
+	switch out.Type {
+	case Sidecar, Ingress, Router:
+	default:
+		return out, fmt.Errorf("invalid node type (valid types: ingress, sidecar, router in the service node %q", s)
+	}
 	out.IPAddress = parts[1]
+
+	// Does query from ingress or router have to carry valid IP address?
+	if net.ParseIP(out.IPAddress) == nil && out.Type == Sidecar {
+		return out, fmt.Errorf("invalid IP address %q in the service node %q", out.IPAddress, s)
+	}
+
```

Comment:
Since you return out, maybe you can move the check after setting out.ID and Domain ( so out can have all the info ). Not sure how the caller is handling the result. 

Long term - the first part (sidecar/ingress/etc) may be more flexible, I don't know how it interacts with the new gateway API but expect 'ingress' to be replaced by the actual ingress/gateway name.

Reply:
Good suggestion. Error is handled properly by the callers by returning HTTP error responses